### PR TITLE
feature 641 MTD input to ExtractTiles

### LINKED ISSUE #641: New Use Case: Feature Relative using MTD output for feature centroid lat/lon
## Describe the Enhancement ##
Currently the Feature Relative use-cases identify the location of the feature from output from the GFDL tracker.  The SBU and UW-CIMSS projects would also like to have support for MTD centroid lat/lon.  I think we need to discuss if this can be handled in the python script that provides lat/lons to the script that cuts out the tile prior to passing into Series-Analysis, or if it requires modifications  to TC-pairs/TC-stat to allow for further refinement.    

### Time Estimate ###
*Estimate the amount of work required here.*
*Issues should represent approximately 1 to 3 days of work.*

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
December 2020

### Funding Source ###

7791041

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required
- [x] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Review **projects** and select relevant **Repository** and **Organization** ones or add "alert:NEED PROJECT ASSIGNMENT" label
- [x] Select **milestone** to next major version milestone or "Future Versions"

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## Enhancement Checklist ##
See the [METplus Workflow](https://dtcenter.github.io/METplus/Contributors_Guide/github_workflow.html) for details.
- [x] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [x] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [x] Complete the development and test your changes.
- [x] Add/update log messages for easier debugging.
- [x] Add/update unit tests.
- [x] Add/update documentation.
- [x] Push local changes to GitHub.
- [x] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [x] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)**, **Project(s)**, **Milestone**, and **Linked issues**
- [x] Iterate until the reviewer(s) accept and merge your changes.
- [x] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by georgemccabe**
Refactor ExtractTiles wrapper to allow MTD data as input to determine storm track

## Pull Request Testing ##

- [X] Describe testing already performed for these changes:</br>

I ran the use case on kiowa

INPUT_BASE:
/d1/projects/METplus/METplus_Data/development/feature_641_mtd_extract_tiles

OUTPUT_BASE:
/d1/personal/mccabe/out-mtd

Output files:
/d1/personal/mccabe/out-mtd/met_tool_wrapper/ExtractTiles/20050807_00

Log file:
/d1/personal/mccabe/out-mtd/logs/metplus.log.20210607160510

- [X] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

Review output from use case run and determine if new logic is sufficient or if additional enhancements are necessary.

- [X] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**

- [X] Do these changes include sufficient testing updates? **[Yes]**

- [X] Will this PR result in changes to the test suite? **[Yes]**</br>
Output from new use case will be added, so updating develop-ref branch is necessary after merging this PR.

- [X] Please complete this pull request review by **7/2/2021**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://dtcenter.github.io/METplus/Contributors_Guide/github_workflow.html) for details.
- [X] Complete the PR definition above.
- [X] Ensure the PR title matches the feature or bugfix branch name.
- [X] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [x] After submitting the PR, select **Linked issues** with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

**Comment by DanielAdriaansen** [comment]
Is the correct link to view the documentation for this use case here:
https://metplus.readthedocs.io/en/feature_641_mtd_extract_tiles/

**Comment by georgemccabe** [comment]
> 
> 
> Is the correct link to view the documentation for this use case here:
> [metplus.readthedocs.io/en/feature_641_mtd_extract_tiles](https://metplus.readthedocs.io/en/feature_641_mtd_extract_tiles/)

Yes

**Review by DanielAdriaansen — CHANGES_REQUESTED** [review]
Only one ExtractTiles use case shows in the docs now, but I think this should be a new one?
https://metplus.readthedocs.io/en/feature_641_mtd_extract_tiles/generated/met_tool_wrapper/index.html#extracttiles

I think there needs to be a new RST/PY file added at:
docs/use_cases/met_tool_wrapper/ExtractTiles/ExtractTiles_mtd.py

**Comment by georgemccabe** [comment]
> 
> 
> Only one ExtractTiles use case shows in the docs now, but I think this should be a new one?
> [metplus.readthedocs.io/en/feature_641_mtd_extract_tiles/generated/met_tool_wrapper/index.html#extracttiles](https://metplus.readthedocs.io/en/feature_641_mtd_extract_tiles/generated/met_tool_wrapper/index.html#extracttiles)
> 
> I think there needs to be a new RST/PY file added at:
> docs/use_cases/met_tool_wrapper/ExtractTiles/ExtractTiles_mtd.py

Yes, you are correct that I forgot to create one. I will do that now.

**Comment by DanielAdriaansen** [comment]
Some additional comments:
EDIT: removed item 1 since it appears to have been done and I missed it (adding new conf items to this section: https://metplus.readthedocs.io/en/feature_641_mtd_extract_tiles/Users_Guide/wrappers.html#id8)
2. Could you run the use-case again with a higher debug verbosity? I was curious to read through the log file and see if it was printing what info it was reading from the MTD output file but I didn't see that. If there's not a level that passes through that info, then let me know.
3. I am confused on what this PR is for- reading #641, it appears the goal is to enhance some feature-relative use cases and John's [comment](https://github.com/dtcenter/METplus/issues/641#issuecomment-854205591) mentions Series Analysis but this PR involves just simply adding a new use case to demonstrate using ExtractTiles wrapper to pull out the centroid lat/lon. Should there be (or maybe there is already?) new issues to update existing use cases to use this new functionality in the ExtractTiles wrapper?

**Comment by georgemccabe** [comment]
> 2. Could you run the use-case again with a higher debug verbosity? I was curious to read through the log file and see if it was printing what info it was reading from the MTD output file but I didn't see that. If there's not a level that passes through that info, then let me know.

The wrapper pulls the centroid lat/lon values from the MTD objects. For each lat/lon point, it subtracts the value of EXTRACT_TILES_LAT_ADJ and EXTRACT_TILES_LON_ADJ, then rounds to the nearest 0.5. It uses this result as the lat0 and lon0 value of the grid definition to create the tiles using RegridDataPlane. The lines that look like this:

DEBUG: FCST lat: 40.51 => 25.5, lon: -99.06 => -114.0

Show how the lat/lon values are converted into the lat0/lon0 values. It follows this format:

DEBUG: <data_type> lat: <lat> => <lat0>, lon: <lon> => <lon0>

The grid definitions that are used in the calls to regrid_data_plane can be see in the commands (highlighted in bold):

INFO: COMMAND: /usr/local/met/bin/regrid_data_plane -v 2 -method NEAREST -width 1 -field 'name="APCP"; level="A03";' -name APCP /d1/projects/METplus/METplus_Data/development/feature_641_mtd_extract_tiles/met_test/data/sample_fcst/2005080700/wrfprs_ruc13_06.tm00_G212 **"latlon 60 60 25.5 -114.0 0.5 0.5"** /d1/personal/mccabe/out-mtd/met_tool_wrapper/ExtractTiles/20050807_00/FCST_TILE_F006_wrfprs_20050807_0000_006.nc

Let me know if there is more information you'd like to see in the logs or if the logs could be modified to be more clear of what it is showing.


> 3. I am confused on what this PR is for- reading #641, it appears the goal is to enhance some feature-relative use cases and John's [comment](https://github.com/dtcenter/METplus/issues/641#issuecomment-854205591) mentions Series Analysis but this PR involves just simply adding a new use case to demonstrate using ExtractTiles wrapper to pull out the centroid lat/lon. Should there be (or maybe there is already?) new issues to update existing use cases to use this new functionality in the ExtractTiles wrapper?

ExtractTiles is used in the feature relative use cases to cut out the tiles of the forecast and observation data along the storm track. These tiles are used as input to SeriesAnalysis. The use case example I created simply demonstrates the step of creating the tiles. Once we have confirmed that this new approach to creating tiles works as expected, we can create a use case that uses MTD data as part of a full feature relative use case. I see this PR as a portion of the work needed to complete issue #641.

**Comment by DanielAdriaansen** [comment]
Thanks George. Would you object to this change to the DEBUG statement:
`DEBUG: <data_type> lat: => , lon: =>`

`DEBUG: <data_type> MTD centroid lat: <val> => tile corner lat: <val> , MTD centroid lon: <val> => tile corner lon: <val>`

The only thing I'm not sure about is whether it makes sense to include "MTD centroid", since I suppose originally this was using a different type of lat/lon to start with. If that's the case, maybe something like:
`DEBUG: <data_type> lat: <val> => tile corner lat: <val> , lon: <val> => tile corner lon: <val>`?

Also, in the MTD input file, it looks like there are three objects (F001, O001 - F003, O003) but only one "centroid" entry (CF001, CO001). I don't know enough about MTD, but is that expected?

> I see this PR as a portion of the work needed to complete issue #641.

Agreed. Maybe @TaraJensen can help define 1 or 2 new issues that will kick off after this is complete? I'd also advocate for this PR not _closing_ #641 if there needs to be follow-on work, unless we want to close #641 and reference the new issues that are created?

**Comment by georgemccabe** [comment]
@DanielAdriaansen the same function is used if the input lat/lon values are from MTD or TCStat output, but I can modify the log message to be more clear as to what the values represent.

Regarding a single CF/CO entry, see the MET User's Guide section about MTD output:
https://met.readthedocs.io/en/latest/Users_Guide/mode-td.html#mtd-output

> The OBJECT_ID column indicates the single object or pair of objects being described in that line. The OBJECT_CAT column indicates the cluster or pair of clusters to which these object(s) belong. A simple object that is not part of a cluster is assigned a cluster number of zero (as in CF000 or CO000). When pairs of objects belong to the same matching cluster, the OBJECT_CAT column indicates the matching cluster number (as in CF001_CO001). When they do not, the OBJECT_CAT column is set to CF000_CO000.

I agree that we should not close the issue after this PR, but we should make a note in the comments for that issue that this step was completed.

**Comment by DanielAdriaansen** [comment]
OK got it. So based on the MET docs:

> In these columns, the F and O prefixes are used to indicate simple forecast and observation objects, respectively. Similarly, the CF and CO prefixes indicate cluster forecast and observation objects, respectively.

In the sample input file there are three "simple forecast/obs objects" and one "cluster forecast/obs objects", and this modification will only read "cluster" forecast/obs centroid lat/lons but not simple objects?

**Comment by georgemccabe** [comment]
> 
> 
> OK got it. So based on the MET docs:
> 
> > In these columns, the F and O prefixes are used to indicate simple forecast and observation objects, respectively. Similarly, the CF and CO prefixes indicate cluster forecast and observation objects, respectively.
> 
> In the sample input file there are three "simple forecast/obs objects" and one "cluster forecast/obs objects", and this modification will only read "cluster" forecast/obs centroid lat/lons but not simple objects?

Yes, I followed guidance from this comment on the GitHub issue to implement the functionality: https://github.com/dtcenter/METplus/issues/641#issuecomment-854205591

File: docs/use_cases/met_tool_wrapper/ExtractTiles/ExtractTiles_mtd.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+met_tool_wrapper/ExtractTiles/ExtractTiles_mtd.conf
+
+"""
+##############################################################################
+# Scientific Objective
+# --------------------
+#
+# Read a MODE Time Domain (MTD) output file and use the centroid latitude
+# and longitude values of the object pairs to
```
Comment (DanielAdriaansen) [inline docs/use_cases/met_tool_wrapper/ExtractTiles/ExtractTiles_mtd.py]: Maybe we could add "MTD cluster" ahead of "object pairs", and then refer to the MET MTD output docs here:
https://met.readthedocs.io/en/latest/Users_Guide/mode-td.html#mtd-output

I can make these changes if you want.

**Review by DanielAdriaansen — CHANGES_REQUESTED** [review]
One more suggestion.

File: docs/Users_Guide/wrappers.rst [hunk 2]
```diff
@@ -867,12 +867,13 @@ Description
 -----------
 
 The ExtractTiles wrapper is used to regrid and extract subregions from
-paired tropical cyclone tracks that are created by the tc_pairs_wrapper.
+paired tropical cyclone tracks. The tracks are created from output from
+TCStat or MTD.
```
Comment (DanielAdriaansen) [inline docs/Users_Guide/wrappers.rst]: Should we also add something here specifying it only reads cluster object pairs?

**Review by DanielAdriaansen — CHANGES_REQUESTED** [review]
Maybe one more suggestion.

File: docs/Users_Guide/wrappers.rst (hunk 2 above)
Comment (georgemccabe) [reply docs/Users_Guide/wrappers.rst]: Sure, feel free to make any adjustments to the documentation to make things more clear.

**Review by DanielAdriaansen — APPROVED** [review]
I approve this PR.

**Pull request merged**